File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/widget/TimetableWidgetService.java
package de.tum.in.tumcampusapp.component.tumui.calendar.widget;

import android.annotation.SuppressLint;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.widget.RemoteViews;
import android.widget.RemoteViewsService;

import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import de.tum.in.tumcampusapp.R;
import de.tum.in.tumcampusapp.component.tumui.calendar.CalendarController;
import de.tum.in.tumcampusapp.component.tumui.calendar.IntegratedCalendarEvent;
import de.tum.in.tumcampusapp.utils.Const;
import de.tum.in.tumcampusapp.utils.DateTimeUtils;

@SuppressLint("Registered")
public class TimetableWidgetService extends RemoteViewsService {

    @Override
    public RemoteViewsFactory onGetViewFactory(Intent intent) {
        return new TimetableRemoteViewFactory(this.getApplicationContext(), intent);
    }

    private class TimetableRemoteViewFactory implements RemoteViewsService.RemoteViewsFactory {

        private final Context applicationContext;
        private final int appWidgetID;
        private List<IntegratedCalendarEvent> calendarEvents;

        TimetableRemoteViewFactory(Context applicationContext, Intent intent) {
            this.applicationContext = applicationContext;
            this.appWidgetID = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, -1);
            calendarEvents = new ArrayList<>();
        }

        @Override
        public void onCreate() {
        }

        @Override
        public void onDataSetChanged() {
            // Get events
            CalendarController calendarController = new CalendarController(this.applicationContext);
            calendarEvents = calendarController.getNextDaysFromDb(14, this.appWidgetID);

            // Set isFirstOnDay flags
            if (!calendarEvents.isEmpty()) {
                calendarEvents.get(0).setIsFirstOnDay(true);
            }

            for (int i = 1; i < calendarEvents.size(); i++) {
                IntegratedCalendarEvent lastEvent = calendarEvents.get(i - 1);
                DateTime lastTime = new DateTime(lastEvent.getStartTime()
                                                          .getTimeInMillis());

                IntegratedCalendarEvent thisEvent = calendarEvents.get(i);
                DateTime thisTime = new DateTime(thisEvent.getStartTime()
                                                          .getTimeInMillis());

                if (!DateTimeUtils.INSTANCE.isSameDay(lastTime, thisTime)) {
                    thisEvent.setIsFirstOnDay(true);
                }
            }
        }

        @Override
        public void onDestroy() {
        }

        @Override
        public int getCount() {
            return this.calendarEvents.size();
        }

        @Override
        public RemoteViews getViewAt(int position) {
            RemoteViews rv = new RemoteViews(applicationContext.getPackageName(), R.layout.timetable_widget_item);
            if (this.calendarEvents == null) {
                return rv;
            }

            // Get the lecture for this view
            IntegratedCalendarEvent currentItem = this.calendarEvents.get(position);
            if (currentItem == null) {
                return null;
            }

            DateTime startTime = new DateTime(currentItem.getStartTime().getTimeInMillis());

            // Setup the date
            if (currentItem.isFirstOnDay()) {
                rv.setTextViewText(R.id.timetable_widget_date_day, String.valueOf(startTime.getDayOfMonth()));
                rv.setTextViewText(R.id.timetable_widget_date_weekday, startTime.dayOfWeek()
                                                                                .getAsShortText(Locale.getDefault()));
                rv.setViewPadding(R.id.timetable_widget_item, 0, 15, 0, 0);
            } else {
                // Overwrite unused parameters, as the elements are reused they may could be filled with old parameters
                rv.setTextViewText(R.id.timetable_widget_date_day, "");
                rv.setTextViewText(R.id.timetable_widget_date_weekday, "");
                rv.setViewPadding(R.id.timetable_widget_item, 0, 0, 0, 0);
            }

            // TODO: Display month label if event is the first event in a new month

            // Setup event color
            rv.setInt(R.id.timetable_widget_event, "setBackgroundColor", currentItem.getColor());

            // Setup event title
            rv.setTextViewText(R.id.timetable_widget_event_title, currentItem.getName());

            // Setup event time
            DateTimeFormatter formatter = DateTimeFormat.shortTime();

            String startTimeText = formatter.print(startTime);
            DateTime endTime = new DateTime(currentItem.getEndTime().getTimeInMillis());
            String endTimeText = formatter.print(endTime);

            String eventTime =
                    getString(R.string.event_start_end_format_string, startTimeText, endTimeText);
            rv.setTextViewText(R.id.timetable_widget_event_time, eventTime);

            // Setup event location
            rv.setTextViewText(R.id.timetable_widget_event_location, currentItem.getLocation());

            // Setup action to open calendar
            Intent fillInIntent = new Intent();
            fillInIntent.putExtra(Const.EVENT_TIME, currentItem.getStartTime().getTimeInMillis());
            rv.setOnClickFillInIntent(R.id.timetable_widget_event, fillInIntent);

            return rv;
        }

        @Override
        public RemoteViews getLoadingView() {
            return null;
        }

        @Override
        public int getViewTypeCount() {
            return 1;
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        @Override
        public boolean hasStableIds() {
            return true;
        }

    }
}

File: app/src/main/java/de/tum/in/tumcampusapp/component/tumui/calendar/widget/TimetableWidgetService.kt
package de.tum.`in`.tumcampusapp.component.tumui.calendar.widget

import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.widget.RemoteViews
import android.widget.RemoteViewsService
import de.tum.`in`.tumcampusapp.R
import de.tum.`in`.tumcampusapp.component.tumui.calendar.CalendarController
import de.tum.`in`.tumcampusapp.component.tumui.calendar.WidgetCalendarItem
import de.tum.`in`.tumcampusapp.utils.Const
import de.tum.`in`.tumcampusapp.utils.DateTimeUtils
import org.joda.time.DateTime
import org.joda.time.format.DateTimeFormat
import java.util.*
import kotlin.collections.ArrayList

class TimetableWidgetService : RemoteViewsService() {

    override fun onGetViewFactory(intent: Intent): RemoteViewsFactory {
        return TimetableRemoteViewFactory(this.applicationContext, intent)
    }

    private class TimetableRemoteViewFactory internal constructor(
        private val applicationContext: Context,
        intent: Intent
    ) : RemoteViewsFactory {

        private val appWidgetID: Int = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, -1)
        private var calendarEvents: List<WidgetCalendarItem> = ArrayList()

        override fun onCreate() {}

        override fun onDataSetChanged() {
            // Get events
            val calendarController = CalendarController(this.applicationContext)
            calendarEvents = calendarController.getNextDaysFromDb(14, this.appWidgetID)

            // Set isFirstOnDay flags
            if (calendarEvents.isNotEmpty()) {
                calendarEvents[0].isFirstOnDay = true
            }

            for (index in 1 until calendarEvents.size) {
                val (_, _, startTime) = calendarEvents[index - 1]
                val lastTime = DateTime(startTime.millis)

                val thisEvent = calendarEvents[index]
                val thisTime = DateTime(thisEvent.startTime.millis)

                if (!DateTimeUtils.isSameDay(lastTime, thisTime)) {
                    thisEvent.isFirstOnDay = true
                }
            }
        }

        override fun onDestroy() {}

        override fun getCount() = calendarEvents.size

        override fun getViewAt(position: Int): RemoteViews? {
            if (position >= calendarEvents.size) {
                return null
            }

            val remoteViews = RemoteViews(applicationContext.packageName, R.layout.timetable_widget_item)

            // Get the lecture for this view
            val currentItem = this.calendarEvents[position]
            val startTime = DateTime(currentItem.startTime.millis)

            // Setup the date
            if (currentItem.isFirstOnDay) {
                remoteViews.setTextViewText(R.id.timetable_widget_date_day, startTime.dayOfMonth.toString())
                remoteViews.setTextViewText(R.id.timetable_widget_date_weekday, startTime.dayOfWeek()
                        .getAsShortText(Locale.getDefault()))
                remoteViews.setViewPadding(R.id.timetable_widget_item, 0, 15, 0, 0)
            } else {
                // Overwrite unused parameters, as the elements are reused they may could be filled with old parameters
                remoteViews.setTextViewText(R.id.timetable_widget_date_day, "")
                remoteViews.setTextViewText(R.id.timetable_widget_date_weekday, "")
                remoteViews.setViewPadding(R.id.timetable_widget_item, 0, 0, 0, 0)
            }

            // TODO: Display month label if event is the first event in a new month

            // Setup event color
            remoteViews.setInt(R.id.timetable_widget_event, "setBackgroundColor", currentItem.color)

            // Setup event title
            remoteViews.setTextViewText(R.id.timetable_widget_event_title, currentItem.title)

            // Setup event time
            val formatter = DateTimeFormat.shortTime()
            val startTimeText = formatter.print(startTime)
            val endTime = DateTime(currentItem.endTime.millis)
            val endTimeText = formatter.print(endTime)
            val eventTime = applicationContext.getString(R.string.event_start_end_format_string, startTimeText, endTimeText)
            remoteViews.setTextViewText(R.id.timetable_widget_event_time, eventTime)

            // Setup event location
            remoteViews.setTextViewText(R.id.timetable_widget_event_location, currentItem.location)

            // Setup action to open calendar
            val fillInIntent = Intent().apply {
                putExtra(Const.EVENT_TIME, currentItem.startTime.millis)
            }
            remoteViews.setOnClickFillInIntent(R.id.timetable_widget_event, fillInIntent)

            return remoteViews
        }

        override fun getLoadingView() = null

        override fun getViewTypeCount() = 1

        override fun getItemId(position: Int) = position.toLong()

        override fun hasStableIds() = true
    }
}
